### PR TITLE
remove middleware Feature concept

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

sibling: https://github.com/awslabs/smithy-kotlin/pull/536

## Description of changes

Updates middleware and codegen to implement upstream interface changes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-runtime/aws-http/common/src/aws/sdk/kotlin/runtime/http/middleware/UserAgent.kt [hunk 1]
```diff
@@ -22,43 +21,30 @@ internal const val USER_AGENT: String = "User-Agent"
...
+) : ModifyRequestMiddleware {
 
-    public class Config {
-        /**
-         * Metadata that doesn't change per/request (e.g. sdk and environment related metadata)
-         */
-        public var staticMetadata: AwsUserAgentMetadata? = null
+    override fun install(op: SdkHttpOperation<*, *>) {
+        op.execution.mutate.register(this, Phase.Order.After)
     }
```
Comment (ianbotsf) [inline aws-runtime/aws-http/common/src/aws/sdk/kotlin/runtime/http/middleware/UserAgent.kt:32]: **Question**: Isn't `Order.After` the default? How is this different from the default implementation of `install`?

**Review by ianbotsf — APPROVED** [review]
Fantastic!

File: aws-runtime/aws-http/common/src/aws/sdk/kotlin/runtime/http/middleware/UserAgent.kt (hunk 1 above)
Comment (aajtodd) [reply aws-runtime/aws-http/common/src/aws/sdk/kotlin/runtime/http/middleware/UserAgent.kt:32]: I believe it is yes. Probably just a miss, will double check. Good catch

File: codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/protocols/middleware/MutateHeadersMiddleware.kt [hunk 2]
```diff
@@ -16,19 +19,23 @@ class MutateHeadersMiddleware(
...
-            writer.write("set(#S, #S)", it.key, it.value)
-        }
-        extraHeaders.forEach {
-            writer.write("append(#S, #S)", it.key, it.value)
-        }
-        addMissingHeaders.forEach {
-            writer.write("setIfMissing(#S, #S)", it.key, it.value)
+        writer.withBlock("op.install(", ")") {
+            withBlock("#T().apply {", "}", RuntimeTypes.Http.Middlware.MutateHeadersMiddleware) {
+                overrideHeaders.forEach {
```
Comment (kggilmer) [inline codegen/smithy-aws-kotlin-codegen/src/main/kotlin/aws/sdk/kotlin/codegen/protocols/middleware/MutateHeadersMiddleware.kt:29]: nit:

I believe you can omit the `writer.` as your `this` is the writer via `withBlock()`

**Review by kggilmer — APPROVED** [review]
again, love the reduction here.

**Pull request merged**